### ISSUE TITLE
We cannot upgrade to Twisted 15.2.0, because the new logging system already pre-formats the messages

### ISSUE THREAD

**Issue by cyli** [rackerlabs/otter#1580]
We see errors such as:

```
2015-07-09_21:40:30.86939 {
2015-07-09_21:40:30.86940   "@timestamp": "2015-07-09T21:40:30.866551", 
2015-07-09_21:40:30.86940   "@version": 1, 
2015-07-09_21:40:30.86941   "cloud_feed": true, 
2015-07-09_21:40:30.86941   "converger_run_id": "1f0f4358-2683-11e5-ade9-000c29b8bc05", 
2015-07-09_21:40:30.86941   "format": "%(log_legacy)s", 
2015-07-09_21:40:30.86941   "host": "vm", 
2015-07-09_21:40:30.86942   "level": 6, 
2015-07-09_21:40:30.86942   "log_format": "{log_text}", 
2015-07-09_21:40:30.86942   "log_legacy": "<twisted.logger._stdlib.StringifiableFromEvent object at 0x41f7050>", 
2015-07-09_21:40:30.86943   "log_level": "<LogLevel=info>", 
2015-07-09_21:40:30.86943   "log_namespace": "log_legacy", 
2015-07-09_21:40:30.86943   "log_system": "otter", 
2015-07-09_21:40:30.86943   "log_text": "convergence-create-servers", 
2015-07-09_21:40:30.86944   "log_time": 1436478030.866551, 
2015-07-09_21:40:30.86944   "message": "Creating 1 with config {'server': pmap({u'flavorRef': u'performance1-1', u'name': u'A', u'imageRef': u'aab63bcf-89aa-440f-b0c7-c7a1c611914b'
, u'metadata': pmap({'rax:autoscale:group:id': u'ddfb3217-5092-49c9-94a2-142c855d5887', u'create_server_failure': u'{\"message\": \"nope\", \"code\": 400}', 'rax:auto_scaling_group
_id': u'ddfb3217-5092-49c9-94a2-142c855d5887'})})}", 
2015-07-09_21:40:30.86944   "message_formatting_error": "[Failure instance: Traceback: <type 'exceptions.KeyError'>: \"'server'\"\n/opt/otter/.ve/local/lib/python2.7/site-packages/twisted/logger/_observer.py:131:__call__\n/opt/otter/.ve/local/lib/python2.7/site-packages/twisted/logger/_legacy.py:93:__call__\n/opt/otter/otter/log/formatters.py:376:emit\n/opt/otter/otter/log/spec.py:138:validating_observer\n--- <exception caught here> ---\n/opt/otter/otter/log/formatters.py:141:PEP3101FormattingObserver\n]", 
2015-07-09_21:40:30.86946   "num_servers": 1, 
2015-07-09_21:40:30.86946   "otter_facility": "otter", 
2015-07-09_21:40:30.86946   "otter_msg_type": "convergence-create-servers", 
2015-07-09_21:40:30.86946   "otter_service": "converger", 
2015-07-09_21:40:30.86947   "scaling_group_id": "ddfb3217-5092-49c9-94a2-142c855d5887", 
2015-07-09_21:40:30.86947   "server_config": {
2015-07-09_21:40:30.86947     "server": "pmap({u'flavorRef': u'performance1-1', u'name': u'A', u'imageRef': u'aab63bcf-89aa-440f-b0c7-c7a1c611914b', u'metadata': pmap({'rax:autoscale:group:id': u'ddfb3217-5092-49c9-94a2-142c855d5887', u'create_server_failure': u'{\"message\": \"nope\", \"code\": 400}', 'rax:auto_scaling_group_id': u'ddfb3217-5092-49c9-94a2-142c855d5887'})})"
2015-07-09_21:40:30.86948   }, 
```

because `otter.log.formatters.PEP3103FormattingWrapper` calls `.format(**eventDict)` on the "message" and the "why" field, but before it can do that, the twisted logging system has already gotten to it and done the same thing.

This causes a problem when JSON or dictionaries get embedded in the message, because then calling `.format` a second time raises `KeyErrors`.

**Comment by glyph** [comment]
So it looks like otter consistently uses `BoundLog` and `otter.log` throughout the code, so we could fix this in one place on the emitter's side.  Otter looks like it has a different idea of the contract of `log.msg` anyway, one that was already subtly broken (although not quite so visibly so) on older versions of Twisted, hence the somewhat ad-hoc `NON_PEP3101_SYSTEMS` variable.  Twisted expects `message` to be a tuple of strings with no semantic meaning, whereas Otter expects `message` to be a tuple of PEP-3101 strings.  The legacy observer formatter code in the new logger performs a formatting transformation which is a valid application of Twisted's expectation of that value but not Otter's.

**Comment by glyph** [comment]
The problem is that, somehow, `message` is getting formatted via PEP 3101 twice.  This doesn't really make sense to me yet, because as I recall, `twisted.logger` does not consider `message` to be a PEP-3101 string at all, and tries to treat it as having semantic information, specifically to avoid this kind of bug.  I am working on crafting a minimal reproducer.

**Comment by glyph** [comment]
It seems that the cloudfeeds observer is likely to be the culprit.  In `makeService`, it calls `addObserver`, which would put the value returned by `get_cf_observer` *first* in the list of log observers.

What's not clear to me is why this isn't happening with the current system.  I suspect that it *is* but that the failure is somehow being lost since the legacy logging system was somewhat less rigorously careful in error cases.  Nothing about this ordering has changed; the semantics of `addObserver` are roughly the same.  The new system supports debugging this by way of the `log_trace` key, but the old system doesn't, so it's not possible to trivially look at the difference in order of observer execution.

**Comment by cyli** [comment]
In `PEP3101FormattingWrapper`, I wrote the `eventDict` and `eventDict` object ID to a file if it was a cloud feeds event.

*In Twisted 15.0.0*:  The IDs were different (140271618733168 vs 140271618792496), and the dicts were the same - both looked like:

```
{'otter_msg_type': 'convergence-create-servers', 'converger_run_id': '6212ada4-26a9-11e5-9836-000c29b8bc05', 'tenant_id': u'000000', 'system': 'otter', 'scaling_group_id': u'7f9736a8-291a-4a31-abfe-f265fc6cdf71', 'num_servers': 1, 'server_config': {'server': pmap({u'flavorRef': u'performance1-1', u'name': u'A', u'imageRef': u'aab63bcf-89aa-440f-b0c7-c7a1c611914b', u'metadata': pmap({'rax:autoscale:group:id': u'7f9736a8-291a-4a31-abfe-f265fc6cdf71', u'create_server_failure': u'{"message": "nope", "code": 400}', 'rax:auto_scaling_group_id': u'7f9736a8-291a-4a31-abfe-f265fc6cdf71'})})}, 'time': 1436494464.180023, 'message': ('Creating {num_servers} with config {server_config}',), 'cloud_feed': True, 'isError': 0, 'otter_service': 'converger'}
```

*In Twisted 15.2.1*:  The IDs were different (140648569292064 vs 78347376), and the dicts were different.  In the same order as the IDs, the dicts looked like:

```
{'otter_msg_type': 'convergence-create-servers', 'log_namespace': u'log_legacy', 'log_time': 1436494715.288032, 'converger_run_id': 'f7bea07e-26a9-11e5-86de-000c29b8bc05', 'log_system': 'otter', 'tenant_id': u'000000', 'format': '%(log_legacy)s', 'log_level': <LogLevel=info>, 'system': 'otter', 'scaling_group_id': u'25dd984c-c077-4075-9c6a-f6869a3c0aca', 'num_servers': 1, 'server_config': {'server': pmap({u'flavorRef': u'performance1-1', u'name': u'A', u'imageRef': u'aab63bcf-89aa-440f-b0c7-c7a1c611914b', u'metadata': pmap({'rax:autoscale:group:id': u'25dd984c-c077-4075-9c6a-f6869a3c0aca', u'create_server_failure': u'{"message": "nope", "code": 400}', 'rax:auto_scaling_group_id': u'25dd984c-c077-4075-9c6a-f6869a3c0aca'})})}, 'time': 1436494715.288032, 'log_text': 'convergence-create-servers', 'log_format': u'{log_text}', 'message': ('Creating {num_servers} with config {server_config}',), 'cloud_feed': True, 'isError': 0, 'log_legacy': <twisted.logger._stdlib.StringifiableFromEvent object at 0x4a6fa90>, 'otter_service': 'converger'}
```
and 

```
{'format': '%(log_legacy)s', 'otter_msg_type': 'convergence-create-servers', 'scaling_group_id': u'25dd984c-c077-4075-9c6a-f6869a3c0aca', 'num_servers': 1, 'log_text': 'convergence-create-servers', 'log_legacy': <twisted.logger._stdlib.StringifiableFromEvent object at 0x4a6fa90>, 'cloud_feed': True, 'log_time': 1436494715.288032, 'otter_service': 'converger', 'log_namespace': u'log_legacy', 'log_level': <LogLevel=info>, 'converger_run_id': 'f7bea07e-26a9-11e5-86de-000c29b8bc05', 'log_system': 'otter', 'tenant_id': u'000000', 'message': ('Creating 1 with config {\'server\': pmap({u\'flavorRef\': u\'performance1-1\', u\'name\': u\'A\', u\'imageRef\': u\'aab63bcf-89aa-440f-b0c7-c7a1c611914b\', u\'metadata\': pmap({\'rax:autoscale:group:id\': u\'25dd984c-c077-4075-9c6a-f6869a3c0aca\', u\'create_server_failure\': u\'{"message": "nope", "code": 400}\', \'rax:auto_scaling_group_id\': u\'25dd984c-c077-4075-9c6a-f6869a3c0aca\'})})}',), 'level': 6, 'system': 'otter', 'isError': 0, 'server_config': {'server': pmap({u'flavorRef': u'performance1-1', u'name': u'A', u'imageRef': u'aab63bcf-89aa-440f-b0c7-c7a1c611914b', u'metadata': pmap({'rax:autoscale:group:id': u'25dd984c-c077-4075-9c6a-f6869a3c0aca', u'create_server_failure': u'{"message": "nope", "code": 400}', 'rax:auto_scaling_group_id': u'25dd984c-c077-4075-9c6a-f6869a3c0aca'})})}, 'time': 1436494715.288032, 'log_format': u'{log_text}'}
```

**Comment by glyph** [comment]
Aha. `throttling_wrapper` does an `event = event.copy()`.  `throttling_wrapper` is wrapped around `observer_factory` and `observer_factory_debug` but *not* around `get_cf_observer`.  So, if somehow `get_cf_observer` managed to come *first* when using Twisted 15.2.1, that would explain the difference.

**Comment by glyph** [comment]
(I still have no idea how that would happen; it makes no sense.  In both versions, `addObserver` just appends, and `makeService` calls all the service-creation code before starting logging.  But this seems to be a plausible difference to overlook.)

**Comment by cyli** [comment]
I still have no idea how this happens, but the CF observer must be called twice, because when CF is turned on, I see duplicate events being sent (see #1584).

**Comment by glyph** [comment]
The issue is ordering of log messages; CF is not invoked twice.

The main Otter logger is added on the `twistd` command line.  The CF logger is added as part of `makeService`.  In both 15.0.0 and 15.2.1, this means that `addObserver` is called with the CF logger *first*.

However, on Twisted 15.0.0, observers are invoked *in reverse order* during event emission.

https://github.com/twisted/twisted/blob/twisted-15.0.0/twisted/python/log.py#L201

I believe this was done in a sort of misguided attempt to deal with modification of the observer list during iteration.  It's still not completely correct.

The invocation on Twisted 15.2.1 is much more straightforward, and simply invokes observers in order:

https://github.com/twisted/twisted/blob/twisted-15.2.1/twisted/logger/_observer.py#L126

This is still not resistant to modification during iteration, and probably something should be done about that, but at least it's easier to understand.

This means that on 15.2.1, the CF observer is invoked first, and by the time we get around to the main observer, the message has been mutated in an invalid way ("message" is no longer a tuple, it's no longer a format string).  On 15.0.0, the more-polite main observer is invoked first, and it makes a copy of the event rather than mutating the main one.

The best fix for now, I think, would be to decouple the copying of the event message from the throttling observer, and wrap all of Otter's somewhat idiosyncratic log observers in the copy-it-first decorator.

**Comment by glyph** [comment]
(Or rather, maybe CF _is_ invoked twice, but that's not where the message format error is coming from; it's a separate issue.)
